pkg/koordlet/resmanager/cpu_suppress.go

```diff
@@ -158,95 +159,85 @@ func (r *CPUSuppress) calculateBESuppressCPU(node *corev1.Node, nodeMetric *metr
+	for _, processorInfos := range cpuBucketOfNode {
+		cpuBucket = append(cpuBucket, processorInfos)
 	}
 
-	// first we prefer the scattered node/socket, secondly the same core
-	// e.g.
-	// for a processorInfo/prioritizedCPUs list:
-	//	CPUID	CoreID	SocketID	NodeID
-	//	0		0		0			0
-	//	1		1		0			0
-	//	2		2		1			1
-	//	3		3		1			1
-	//	4		0		0			0
-	//	5		1		0			0
-	//	6		2		1			1
-	//	7		3		1			1
-	// it will get sorted as below:
-	//	CPUID	CoreID	SocketID	NodeID
-	//	7		3		1			1
-	//	3		3		1			1
-	//	5		1		0			0
-	//	1		1		0			0
-	//	6		2		1			1
-	//	2		2		1			1
-	//	4		0		0			0
-	//	0		0		0			0
-	sort.Slice(prioritizedCPUs, func(i, j int) bool {
-		a, b := prioritizedCPUs[i], prioritizedCPUs[j]
-		if indexOfNode[a.CoreID] != indexOfNode[b.CoreID] {
-			return indexOfNode[a.CoreID] > indexOfNode[b.CoreID]
-		}
-		if numProcessorsOfNode[getNodeIndex(a)] != numProcessorsOfNode[getNodeIndex(b)] {
-			return numProcessorsOfNode[getNodeIndex(a)] > numProcessorsOfNode[getNodeIndex(b)]
-		}
-		if a.CoreID != b.CoreID {
-			return a.CoreID > b.CoreID
+	for index := range cpuBucket {
+		sort.Slice(cpuBucket[index], func(i, j int) bool {
+			if cpuBucket[index][i].CoreID == cpuBucket[index][j].CoreID {
+				return cpuBucket[index][i].CPUID < cpuBucket[index][j].CPUID
+			}
+			return cpuBucket[index][i].CoreID < cpuBucket[index][j].CoreID
+		})
+	}
+
+	sort.Slice(cpuBucket, func(i, j int) bool {
+		if len(cpuBucket[i]) == len(cpuBucket[j]) {
+			return cpuBucket[i][0].CPUID < cpuBucket[j][0].CPUID
 		}
-		return a.CPUID > b.CPUID
+		return len(cpuBucket[i]) > len(cpuBucket[j])
 	})
 
 	needCPUs := cpus
-	for i := range prioritizedCPUs {
-		if needCPUs <= 0 {
-			break
+	bucketIndex := 0
+	for needCPUs > 0 {
+		if needCPUs == 1 || len(cpuBucket[bucketIndex]) == 1 {
+			// prefer the single core cpu id
+			selectdIndex := 0
+			for i := 0; i < len(cpuBucket[bucketIndex]); i++ {
+				if cpuBucket[bucketIndex][i].CPUID < 0 {
+					continue
+				}
+				if (i == 0 || cpuBucket[bucketIndex][i].CoreID != cpuBucket[bucketIndex][i-1].CoreID) && (i+1 == len(cpuBucket[bucketIndex]) || cpuBucket[bucketIndex][i].CoreID != cpuBucket[bucketIndex][i+1].CoreID) {
+					selectdIndex = i
+					break
+				}
+			}
+			CPUSets = append(CPUSets, cpuBucket[bucketIndex][selectdIndex].CPUID)
+			cpuBucket[bucketIndex][selectdIndex].CPUID = -1
+			needCPUs--
+			bucketIndex = (bucketIndex + 1) % len(cpuBucket)
+			continue
+		}
+		// prefer the same core
+		selectdIndex := 0
+		for i := 0; i < len(cpuBucket[bucketIndex])-1; i++ {
+			if cpuBucket[bucketIndex][i].CPUID < 0 {
+				continue
+			}
+			if cpuBucket[bucketIndex][i].CoreID == cpuBucket[bucketIndex][i+1].CoreID {
+				selectdIndex = i
+				break
+			}
 		}
-		CPUSets = append(CPUSets, prioritizedCPUs[i].CPUID)
-		needCPUs--
+		CPUSets = append(CPUSets, cpuBucket[bucketIndex][selectdIndex].CPUID, cpuBucket[bucketIndex][selectdIndex+1].CPUID)
+		cpuBucket[bucketIndex][selectdIndex].CPUID = -1
```

Comment:
mark "-1" to cpu id seems tricky, consider using a bitmap to record allocation result